pkg/object/mqttproxy/broker.go

```diff
@@ -0,0 +1,348 @@
+/*
+ * Copyright (c) 2017, MegaEase
+ * All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package mqttproxy
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"crypto/tls"
+	"encoding/base64"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"net"
+	"net/http"
+	"sync"
+
+	"github.com/eclipse/paho.mqtt.golang/packets"
+	"github.com/megaease/easegress/pkg/api"
+	"github.com/megaease/easegress/pkg/logger"
+)
+
+type (
+	// Broker is MQTT server, will manage client, topic, session, etc.
+	Broker struct {
+		sync.RWMutex
+		egName string
+		name   string
+		spec   *Spec
+
+		listener   net.Listener
+		backend    backendMQ
+		clients    map[string]*Client
+		sha256Auth map[string]string
+		tlsCfg     *tls.Config
+
+		sessMgr   *SessionManager
+		topicMgr  *TopicManager
+		memberURL func(string, string) ([]string, error)
+
+		// done is the channel for shutdowning this proxy.
+		done chan struct{}
+	}
+
+	// HTTPJsonData is json data received from http endpoint used to send back to clients
+	HTTPJsonData struct {
+		Topic       string `json:"topic"`
+		Qos         int    `json:"qos"`
+		Payload     string `json:"payload"`
+		Base64      bool   `json:"base64"`
+		Distributed bool   `json:"distributed"`
+	}
+)
+
+func sha256Sum(data []byte) string {
+	sha256Bytes := sha256.Sum256(data)
+	return hex.EncodeToString(sha256Bytes[:])
+}
+
+func newBroker(spec *Spec, store storage, memberURL func(string, string) ([]string, error)) *Broker {
+	broker := &Broker{
+		egName:     spec.EGName,
+		name:       spec.Name,
+		spec:       spec,
+		backend:    newBackendMQ(spec),
+		clients:    make(map[string]*Client),
+		sha256Auth: make(map[string]string),
+		memberURL:  memberURL,
+		done:       make(chan struct{}),
+	}
+
+	for _, a := range spec.Auth {
+		passwd, err := base64.StdEncoding.DecodeString(a.PassBase64)
+		if err != nil {
+			logger.Errorf("auth with name <%v>, base64 password <%v> decode failed, err:%v", a.UserName, a.PassBase64, err)
```

Comment:
I think `passwd` checking failure is a fatal error for creating a `Broker`. Should return `nil` or some error here?

Reply:
Well, I do it in this way to make the MQTT still work with one or more available auth, but your idea may be better to fail the code as soon as possible.

Reply:
If so, propose to update it with 
```

   if len(a.PassBase64) != 0 {
          // checking, if failed, fatal
   }
  
   if len(a.AnotherAuth) != 0 {
          // checking, if failed, fatal
   }
```
This looks more like your description above. If users provide password with error contents, then I think we should treat it more strictly.

Reply:
Yeah, although it will make the server start a little bit early, but it may cause bug difficult to debug, thanks a lot for your advice! I will fix them all as soon as possible!